fem/libceed/algebraic.cpp

```diff
@@ -0,0 +1,627 @@
+         op_assembled = RAP(&hypre_local, P);
+      }
+      HypreParMatrix *mat_e = op_assembled->EliminateRowsCols(ess_tdofs);
+      delete mat_e;
+      amg = new HypreBoomerAMG(*op_assembled);
+      amg->SetPrintLevel(0);
+   }
+   void SetOperator(const Operator &op) { }
+   void Mult(const Vector &x, Vector &y) const { amg->Mult(x, y); }
+   ~CeedAMG()
+   {
+      delete op_assembled;
+      delete amg;
+      delete mat_local;
+   }
+private:
+   SparseMatrix *mat_local;
+   HypreParMatrix *op_assembled;
+   HypreBoomerAMG *amg;
+};
+
+#endif
+
+void CoarsenEssentialDofs(const Operator &interp,
+                          const Array<int> &ho_ess_tdofs,
+                          Array<int> &alg_lo_ess_tdofs)
+{
+   Vector ho_boundary_ones(interp.Height());
+   ho_boundary_ones = 0.0;
+   const int *ho_ess_tdofs_h = ho_ess_tdofs.HostRead();
+   for (int i=0; i<ho_ess_tdofs.Size(); ++i)
+   {
+      ho_boundary_ones[ho_ess_tdofs_h[i]] = 1.0;
+   }
+   Vector lo_boundary_ones(interp.Width());
+   interp.MultTranspose(ho_boundary_ones, lo_boundary_ones);
+   auto lobo = lo_boundary_ones.HostRead();
+   for (int i = 0; i < lo_boundary_ones.Size(); ++i)
+   {
+      if (lobo[i] > 0.9)
+      {
+         alg_lo_ess_tdofs.Append(i);
+      }
+   }
+}
+
+void AddToCompositeOperator(BilinearFormIntegrator *integ, CeedOperator op)
+{
+   if (integ->SupportsCeed())
+   {
+      CeedCompositeOperatorAddSub(op, integ->GetCeedData()->oper);
+   }
+   else
+   {
+      MFEM_ABORT("This integrator does not support Ceed!");
+   }
+}
+
+CeedOperator CreateCeedCompositeOperatorFromBilinearForm(BilinearForm &form)
+{
+   int ierr;
+   CeedOperator op;
+   ierr = CeedCompositeOperatorCreate(internal::ceed, &op); PCeedChk(ierr);
+
+   // Get the domain bilinear form integrators (DBFIs)
+   Array<BilinearFormIntegrator*> *bffis = form.GetDBFI();
+   for (int i = 0; i < bffis->Size(); ++i)
+   {
+      AddToCompositeOperator((*bffis)[i], op);
+   }
+   return op;
+}
+
+CeedOperator CoarsenCeedCompositeOperator(
+   CeedOperator op,
+   CeedElemRestriction er,
+   CeedBasis c2f,
+   int order_reduction
+)
+{
+   int ierr;
+   bool isComposite;
+   ierr = CeedOperatorIsComposite(op, &isComposite); PCeedChk(ierr);
+   MFEM_ASSERT(isComposite, "");
+
+   CeedOperator op_coarse;
+   ierr = CeedCompositeOperatorCreate(internal::ceed,
+                                      &op_coarse); PCeedChk(ierr);
+
+   int nsub;
+   ierr = CeedOperatorGetNumSub(op, &nsub); PCeedChk(ierr);
+   CeedOperator *subops;
+   ierr = CeedOperatorGetSubList(op, &subops); PCeedChk(ierr);
+   for (int isub=0; isub<nsub; ++isub)
+   {
+      CeedOperator subop = subops[isub];
+      CeedBasis basis_coarse, basis_c2f;
+      CeedOperator subop_coarse;
+      ierr = CeedATPMGOperator(subop, order_reduction, er, &basis_coarse,
+                               &basis_c2f, &subop_coarse); PCeedChk(ierr);
+      // destructions below make sense because these objects are
```

Comment:
Not sure what you are asking. I added the comment because the objects in question stay around and in fact are used later in the code, so I wanted to remind myself and future maintainers what is going on (ie, we just change the refcount so a later "real" destruction cleanly deletes the objects).